posts/2024-01-31-Performance-Doris_Amoakohene/index.qmd

```diff
@@ -0,0 +1,249 @@
+
+It is an essential practice to deliver a performant and reliable package to end-users.
+
+# **What are the performance tests?**
+Our atime performance tests aims to assess the data.table repository by benchmarking its performance and gathering information on memory and time usage.  
+
+By conducting these tests, we can gain insights into the package's performance efficiency.
+
+## When using atime_versions, there are six main arguments:
+
+1. `pkg.path`: This argument specifies the location on your system where you have stored a git clone of the `data.table` package.
+
+2. `pkg.edit.fun`: The default behavior of pkg.edit.fun is designed to work with Rcpp packages and involves replacing instances of "PKG" with "PKG.SHA" in the package code. Any occurrences of the string "PKG" within the package code will be replaced with "PKG.SHA", where "SHA" represents the commit SHA/ids associated with the version being installed.
+
+3. `N`: This argument determines the number of iterations for the benchmarking process. It is a sequence of numbers that define different data sizes to test the performance of the operation.
+
+4. `setup`: This section contains the setup code for generating the dataset used in the benchmarking process, the setup is determined by the value of `N`. 
+
+5. `expr`: This section contains the expression that represents the operation being benchmarked. It uses the `data.table::`[.data.table`` syntax to perform the  operation on the dataset. 
+
+In the given syntax `` data.table::`[.data.table` ``, the first part `data.table::` installs and loads different versions of the data.table package based on the specified commit IDs. Hence, data.table:: will be translated to data.table.SHA1:: for some version hash SHA1. 
+Following that, the expression specified within `` `[.data.table`` `` is executed on each installed version. This process is repeated for all the specified commit IDs in the code.
+
+For example:
+
+```
+data.table.ec1259af1bf13fc0c96a1d3f9e84d55d8106a9a4:::`[.data.table`(DT, , .(v3=mean(v3, na.rm=TRUE)), by=id3, verbose=TRUE)
+
+```
+
+In this example, the expression `[.data.table` is executed on the `DT` dataset using the specified commit ID (`ec1259af1bf13fc0c96a1d3f9e84d55d8106a9a4`) of the data.table package. The expression calculates the mean of the `v3` column (ignoring missing values) grouped by `id3`, and the `verbose=TRUE` argument enables verbose output during the operation. This process is typically repeated for all commit IDs in your code to compare the performance of different versions of the data.table package.
+
+6. `...` : This section specifies the different versions of the data.table packages that will be tested. It includes three versions: "Before," "Regression," and "Fixed." Each version is associated with a specific commit id.
```

Comment:
I'm not sure if this can be labelled as a 'section', as it looks like an ellipsis (used for varying arguments usually) if not just a placeholder for continuation or incomplete code/output (also where is it used below?)

Reply:
that is how the named versions is defined in the ?atime::atime_versions(), 

```
... : named versions.

```